lib/algos/cache/GapSet.ts

```diff
@@ -0,0 +1,352 @@
+            gap = null;
+        }
+        if (gap) {
+            if (gap.lastKey >= lastKey) {
+                // return fully overlapping gap already cached
+                return { gap, weightToMerge: 0 };
+            }
+            if (gap.lastKey === firstKey) {
+                // new gap is chained with last gap: merge the full weight value
+                weightToMerge = weight;
+            } else if (gap.firstKey === firstKey) {
+                // last gap is fully contained in the new gap: remove the
+                // last gap's weight from the weight to merge
+                weightToMerge = Math.max(weight - gap.weight, 0);
+            } else {
+                // last gap is not fully contained nor chained: default to returning
+                // the new weight (heuristic)
+                weightToMerge = weight;
+            }
+            if (weight > this._maxWeight) {
+                // if the new gap's weight exceeds the max, split the existing gap found in
+                // two chained gaps, by using the same key as a link (this is important
+                // to ensure that lookupGap() can return them as a single contiguous gap)
+                gap = {
+                    firstKey: gap.lastKey,
+                    // split gap is temporarily single-key, will be extended by setGap()
+                    lastKey: gap.lastKey,
+                    // split gap only contains the weight not already included in the previous gap
+                    //weight: Math.max(weight - gap.weight, 0),
+                    weight: 0,
+                };
+                // there may be an existing gap starting with 'lastKey': delete it first
+                this._gaps.delete(gap);
+                // then add the new split gap to be extended thereafter
+                this._gaps.add(gap);
+                return { gap, weightToMerge };
+            }
+        } else {
+            // create a new single-key gap that will be extended to 'lastKey' in setGap()
+            // during the merge process
+            gap = {
+                firstKey,
+                lastKey: firstKey,
+                weight: 0,
+            };
+            this._gaps.add(gap);
+            weightToMerge = weight;
+        }
+        return { gap, weightToMerge };
+    }
+
+    /**
+     * Record a gap between two keys, associated with a weight to limit
+     * individual gap sizes in the cache.
+     *
+     * The function handles splitting and merging existing gaps to
+     * maintain an optimal weight of cache entries.
+     *
+     * @param {string} firstKey - first key of the gap
+     * @param {string} lastKey - last key of the gap, must be greater
+     * or equal than 'firstKey'
+     * @param {number} weight - total weight between 'firstKey' and 'lastKey'
+     * @return {GapSetEntry} - existing or new gap entry
+     */
+    setGap(firstKey: string, lastKey: string, weight: number): GapSetEntry {
```

Comment:
I was thinking of an alternate logic for setGap (still similar though). 

In a for loop: 
- find the first gap in the GapSet that overlaps with the new gap and re-arrange the situation if needed (same as in `_lookupOrCreateGap()`)
- at this point, we have probably found part of the gap already inside the set, or a stretch of new keys that we have inserted, either way the new gap is being reduced
- continue with diminishing new gap until it's empty

I feel like it's a little bit more straightforward to reason about, what do you think?

Reply:
Sure! I'll put the weights aside for now. Basically, I'm thinking that we are adding a `newGap{first, last}` to a `GapSet`, let's call it `gapSet.mergeGap(newGap)`

First locate the first gap in `GapSet` that strictly overlaps with `newGap` (no chaining-style overlap), let's call it `gap1{first1, last1}`. Now there are three cases: 
- `gap1` completely encompasses `newGap`, we're done. 
- `gap1` starts before or on the same key as `newGap`, then we can do `gapSet.mergeGap(gap{last1, last})`
- `gap1` starts after `newGap`, then we insert `gap{first, first1}` and call `gapSet.mergeGap(gap{last1, last})`

I wrote it recursively only because I think it's easier to express. 

What I meant in my comment is that I feel like the handling of these three cases is kind of what you do in `_lookupOrCreateGap()`. There are more cases because I have hidden some of the cases behind wording like "first gap that strictly overlaps". 

I just thought it was easier to reason about this way but it might also just be personal taste, feel free to take it or leave it.